toolkit/tools/internal/pkggraph/pkggraph.go

```diff
@@ -1355,27 +1473,57 @@ func (g *PkgGraph) fixPrebuiltSRPMsCycle(trimmedCycle []*PkgNode) (err error) {
 
 			logger.Log.Debugf("Adding a 'PreBuilt' node '%s' with id %d.", preBuiltNode.FriendlyName(), preBuiltNode.ID())
 
-			parentNodes := g.To(currentNode.ID())
-			for parentNodes.Next() {
-				parentNode := parentNodes.Node().(*PkgNode)
-				if parentNode.Type == TypeLocalBuild && parentNode.SrpmPath == previousNode.SrpmPath {
-					g.RemoveEdge(parentNode.ID(), currentNode.ID())
-
-					err = g.AddEdge(parentNode, preBuiltNode)
-					if err != nil {
-						logger.Log.Errorf("Adding edge failed for %v -> %v", parentNode, preBuiltNode)
-						return
-					}
-				}
-			}
+			g.replaceCurrentRunNodeWithNewNode(currentNode, preBuiltNode, previousNode)
+			logger.Log.Infof("Cycle fixed using prebuilt node: %s with id %d", preBuiltNode.FriendlyName(), preBuiltNode.ID())
 
 			return
 		}
+		if resolveCyclesFromUpstream {
+			var isAvailableUpstream bool
+			isAvailableUpstream, err = findAllRPMSUpstream([]*PkgNode{currentNode}, false, cloner)
+			// Cloner errors are not fatal - may return one when the RPM is not found.
+			if err != nil {
+				logger.Log.Debugf("Error while checking if SRPM is available upstream: %v", err)
+				continue
+			}
+			if isAvailableUpstream {
+				logger.Log.Debugf("Cycle contains SRPM '%s' that is availabe in repo. Replacing edges from build nodes associated with '%s' with an edge to a new 'remote-unresolved' node.", currentNode.SrpmPath, previousNode.SrpmPath)
+				g.cloneAndReplaceRunToRemote(currentNode, previousNode)
+				return
+			}
 
+		}
 		currentNode = previousNode
 	}
 
-	return fmt.Errorf("cycle contains no pre-build SRPMs, unresolvable")
+	if !ignoreVersionToResolveSelfDep || !resolveCyclesFromUpstream {
+		logger.Log.Debugf("Cycle can't be fixed using pre-built SRPMs.")
+		return
+	}
+
+	logger.Log.Debug("Trying to resolve cycle by searching packages in PMC after ignoring package version.")
+	// Still cycle is not resolved. So try to resolve it by ignoring package version
+	currentNode = trimmedCycle[len(trimmedCycle)-1]
+	for _, previousNode := range trimmedCycle {
+		buildToRunEdge := previousNode.Type == TypeLocalBuild && currentNode.Type == TypeLocalRun
+		if !buildToRunEdge {
+			continue
+		}
+		var isAvailableUpstream bool
+		isAvailableUpstream, err = findAllRPMSUpstream([]*PkgNode{currentNode}, true, cloner)
+		// Cloner errors are not fatal - may return one when the RPM is not found.
+		if err != nil {
+			logger.Log.Debugf("Error while checking if SRPM is available upstream: %v", err)
+			continue
+		}
+		if isAvailableUpstream {
+			logger.Log.Debugf("Cycle contains SRPM '%s' that is availabe in repo. Replacing edges from build nodes associated with '%s' with an edge to a new 'remote-unresolved' node.", currentNode.SrpmPath, previousNode.SrpmPath)
+			g.cloneAndReplaceRunToRemote(currentNode, previousNode)
+			return
+		}
+	}
+
+	return fmt.Errorf("Cycle can't be resolved with prebuilt/PMC RPMs. unresolvable")
```

Comment:
We try to follow [Golang's convention](https://github.com/golang/go/wiki/CodeReviewComments#error-strings) for error string and make the `fmt.Error*` strings begin with a small letter.
```suggestion
	return fmt.Errorf("cycle can't be resolved with prebuilt/PMC RPMs. Unresolvable")
```